File: frontend/src/lib/api/feedback.ts
import { apiClient } from "@/lib/axios";

export type FeedbackType =
  | "bug_report"
  | "feature_request"
  | "ai_experience"
  | "job_report"
  | "general";

export type FeedbackStatus = "new" | "in_progress" | "resolved" | "rejected";
export type FeedbackPriority = "low" | "medium" | "high" | "urgent";
export type UserRoleType = "candidate" | "employer" | "guest";

export interface UserFeedback {
  id: number;
  user_id: number | null;
  user_role: UserRoleType;
  sender_name: string;
  sender_email: string;
  sender_phone?: string | null;
  feedback_type: FeedbackType;
  title: string;
  content: string;
  rating?: number | null;
  target_id?: string | null;
  target_type?: string | null;
  status: FeedbackStatus;
  priority: FeedbackPriority;
  admin_notes?: string | null;
  admin_response?: string | null;
  resolved_by?: number | null;
  resolved_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface FeedbackCreatePayload {
  sender_name: string;
  sender_email: string;
  sender_phone?: string;
  feedback_type: FeedbackType;
  title: string;
  content: string;
  rating?: number;
  target_id?: string;
  target_type?: string;
}

export interface FeedbackUpdateAdminPayload {
  status?: FeedbackStatus;
  priority?: FeedbackPriority;
  admin_notes?: string;
  admin_response?: string;
}

export interface FeedbackListResponse {
  items: UserFeedback[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface FeedbackStatsResponse {
  total_feedbacks: number;
  pending_feedbacks: number;
  in_progress_feedbacks: number;
  resolved_feedbacks: number;
  avg_csat_rating?: number | null;
  by_role: Record<string, number>;
  by_type: Record<string, number>;
}

export interface FeedbackFilterParams {
  user_role?: string;
  feedback_type?: string;
  status?: string;
  priority?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

/** Public / User submit feedback */
export async function submitFeedback(payload: FeedbackCreatePayload): Promise<UserFeedback> {
  const { data } = await apiClient.post<UserFeedback>("/feedback", payload);
  return data;
}

/** Admin get feedback list */
export async function getAdminFeedbacks(
  params: FeedbackFilterParams = {}
): Promise<FeedbackListResponse> {
  const { data } = await apiClient.get<FeedbackListResponse>("/admin/feedback", { params });
  return data;
}

/** Admin get feedback stats */
export async function getAdminFeedbackStats(): Promise<FeedbackStatsResponse> {
  const { data } = await apiClient.get<FeedbackStatsResponse>("/admin/feedback/stats");
  return data;
}

/** Admin get feedback detail */
export async function getAdminFeedbackDetail(id: number): Promise<UserFeedback> {
  const { data } = await apiClient.get<UserFeedback>(`/admin/feedback/${id}`);
  return data;
}

/** Admin update feedback */
export async function updateAdminFeedback(
  id: number,
  payload: FeedbackUpdateAdminPayload
): Promise<UserFeedback> {
  const { data } = await apiClient.patch<UserFeedback>(`/admin/feedback/${id}`, payload);
  return data;
}

File: backend/app/schemas/feedback.py
"""Pydantic schemas for User Feedback System."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, EmailStr, Field


class FeedbackCreate(BaseModel):
    """Schema for submitting user feedback (public or authenticated)."""

    sender_name: str = Field(..., min_length=2, max_length=255, description="Full name of sender")
    sender_email: EmailStr = Field(..., description="Contact email address")
    sender_phone: str | None = Field(None, max_length=50, description="Optional contact phone")
    feedback_type: Literal[
        "bug_report", "feature_request", "ai_experience", "job_report", "general"
    ] = Field("general", description="Category of feedback")
    title: str = Field(..., min_length=3, max_length=255, description="Brief summary / title")
    content: str = Field(..., min_length=10, description="Detailed explanation of feedback")
    rating: int | None = Field(None, ge=1, le=5, description="Satisfaction score 1-5 stars")
    target_id: str | None = Field(None, max_length=64, description="Optional ID of target entity")
    target_type: str | None = Field(None, max_length=50, description="Type of target (job, cv...)")


class FeedbackUpdateAdmin(BaseModel):
    """Schema for privileged admin updating feedback status, priority or responding."""

    status: Literal["new", "in_progress", "resolved", "rejected"] | None = None
    priority: Literal["low", "medium", "high", "urgent"] | None = None
    admin_notes: str | None = Field(None, description="Internal admin note")
    admin_response: str | None = Field(None, description="Resolution reply to user")


class FeedbackResponse(BaseModel):
    """Schema for returning feedback details."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int | None = None
    user_role: str
    sender_name: str
    sender_email: str
    sender_phone: str | None = None
    feedback_type: str
    title: str
    content: str
    rating: int | None = None
    target_id: str | None = None
    target_type: str | None = None
    status: str
    priority: str
    admin_notes: str | None = None
    admin_response: str | None = None
    resolved_by: int | None = None
    resolved_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class FeedbackListResponse(BaseModel):
    """Paginated feedback list response."""

    items: list[FeedbackResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class FeedbackStatsResponse(BaseModel):
    """Aggregated feedback statistics for Admin Dashboard."""

    total_feedbacks: int
    pending_feedbacks: int
    in_progress_feedbacks: int
    resolved_feedbacks: int
    avg_csat_rating: float | None = None
    by_role: dict[str, int]
    by_type: dict[str, int]
